### PR TITLE
Update to latest otel

### PR DESCRIPTION
**Pull request by trask**
Includes fix for problematic jdbc wrapper

### CODE REVIEW

**Comment by heyams** [comment]
Is this https://github.com/open-telemetry/opentelemetry-java/issues/4400 fixed in 1.14? Just curious so that I can work on that.

**Comment by trask** [comment]
> Is this [open-telemetry/opentelemetry-java#4381](https://github.com/open-telemetry/opentelemetry-java/issues/4381) fixed in 1.14? Just curious so that I can work on that.

It's not fixed yet, but I think you can start working on it, and we can mitigate the issue in our distro if needed by setting the logging level for `io.opentelemetry.sdk.metrics.internal.state.MetricStorageRegistry` to `ERROR` (so that we won't log those warnings for now)

File: agent/instrumentation/applicationinsights-web-2.3/src/main/java/io/opentelemetry/javaagent/instrumentation/applicationinsightsweb/ApplicationInsightsWebInstrumentationModule.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import static java.util.Arrays.asList;
+
+import com.google.auto.service.AutoService;
+import io.opentelemetry.javaagent.extension.instrumentation.InstrumentationModule;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
+import java.util.List;
+
+@AutoService(InstrumentationModule.class)
+public class ApplicationInsightsWebInstrumentationModule extends InstrumentationModule {
```
Comment (heyams) [inline agent/instrumentation/applicationinsights-web-2.3/src/main/java/io/opentelemetry/javaagent/instrumentation/applicationinsightsweb/ApplicationInsightsWebInstrumentationModule.java:32]: AzureMonitorInstrumentationModule? we used azuremonitor prefix for exporters.

File: agent/instrumentation/applicationinsights-web-2.3/src/main/java/io/opentelemetry/javaagent/instrumentation/applicationinsightsweb/ThreadContextInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import io.opentelemetry.instrumentation.api.field.VirtualField;
+import io.opentelemetry.instrumentation.api.instrumenter.LocalRootSpan;
+import io.opentelemetry.javaagent.bootstrap.Java8BytecodeBridge;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+public class ThreadContextInstrumentation implements TypeInstrumentation {
```
Comment (heyams) [inline agent/instrumentation/applicationinsights-web-2.3/src/main/java/io/opentelemetry/javaagent/instrumentation/applicationinsightsweb/ThreadContextInstrumentation.java:41]: are these subclasses of TypeInstrumentation straight from upstream? why do we need to have our own implementations? I need more context about it.. and don't understand the implication and reason behind this.

File: agent/instrumentation/applicationinsights-web-2.3/src/main/java/io/opentelemetry/javaagent/instrumentation/applicationinsightsweb/ApplicationInsightsWebInstrumentationModule.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package io.opentelemetry.javaagent.instrumentation.applicationinsightsweb;
```
[the remark this answers is not in the document]
Comment (trask) [reply agent/instrumentation/applicationinsights-web-2.3/src/main/java/io/opentelemetry/javaagent/instrumentation/applicationinsightsweb/ApplicationInsightsWebInstrumentationModule.java:22]: this has been convention upstream to concatenate the artifactId being instrumented

File: agent/instrumentation/applicationinsights-web-2.3/src/main/java/io/opentelemetry/javaagent/instrumentation/applicationinsightsweb/ApplicationInsightsWebInstrumentationModule.java (hunk 1 above)
Comment (trask) [reply agent/instrumentation/applicationinsights-web-2.3/src/main/java/io/opentelemetry/javaagent/instrumentation/applicationinsightsweb/ApplicationInsightsWebInstrumentationModule.java:32]: this is instrumentation of the old Application Insights 2.x SDK

File: agent/instrumentation/applicationinsights-web-2.3/src/main/java/io/opentelemetry/javaagent/instrumentation/applicationinsightsweb/ThreadContextInstrumentation.java (hunk 2 above)
Comment (trask) [reply agent/instrumentation/applicationinsights-web-2.3/src/main/java/io/opentelemetry/javaagent/instrumentation/applicationinsightsweb/ThreadContextInstrumentation.java:41]: these are just moved over from the fork (I'm trying to still reduce the fork)

File: agent/agent/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+      exclude("META-INF/INDEX.LIST")
+      exclude("META-INF/*.DSA")
+      exclude("META-INF/*.SF")
+    }
+  }
+  from("${rootProject.projectDir}/NOTICE")
+  from("${rootProject.projectDir}/LICENSE")
+}
+
+// FIXME (trask) publishing
```
Comment (trask) [inline agent/agent/build.gradle.kts:152]: TODO

File: agent/agent-bootstrap/gradle.lockfile [hunk 5]
```diff
@@ -9,23 +9,9 @@ com.google.guava:guava-bom:31.0.1-jre=runtimeClasspath
...
-io.opentelemetry.javaagent:opentelemetry-javaagent-instrumentation-api:1.13.0+ai.patches-alpha=runtimeClasspath
-io.opentelemetry:opentelemetry-api:1.13.0=runtimeClasspath
+io.opentelemetry.instrumentation:opentelemetry-instrumentation-bom-alpha:1.14.0.1-alpha=runtimeClasspath
 io.opentelemetry:opentelemetry-bom-alpha:1.13.0-alpha=runtimeClasspath
 io.opentelemetry:opentelemetry-bom:1.13.0=runtimeClasspath
-io.opentelemetry:opentelemetry-context:1.13.0=runtimeClasspath
-io.opentelemetry:opentelemetry-semconv:1.13.0-alpha=runtimeClasspath
 org.junit:junit-bom:5.7.2=runtimeClasspath
-org.slf4j:slf4j-api:1.7.36=runtimeClasspath
+org.slf4j:slf4j-api:1.7.32=runtimeClasspath
```
[the remark this answers is not in the document]
Comment (trask) [reply agent/agent-bootstrap/gradle.lockfile]: fixed

File: agent/agent/build.gradle.kts (hunk 4 above)
Comment (trask) [reply agent/agent/build.gradle.kts:152]: and attaching artifact to PR build is broken, probably for this reason too

**Review by heyams — APPROVED** [review]

File: .pipelines/pipeline.user.windows.official.yml [hunk 6]
```diff
@@ -16,13 +16,11 @@ build:
...
-            - '**/build/libs/*.jar'
-            - '**/build/libs/*-sources.jar'
-            - '**/build/libs/*-javadoc.jar'
-            - '**/build/libs/*.pom'
-          exclude:
-            - 'buildSrc/**/*'
+            - '~/.m2/repository/com/microsoft/azure/applicationinsights-agent/*/*.jar'
+            - '~/.m2/repository/com/microsoft/azure/applicationinsights-agent/*/*-sources.jar'
+            - '~/.m2/repository/com/microsoft/azure/applicationinsights-agent/*/*-javadoc.jar'
+            - '~/.m2/repository/com/microsoft/azure/applicationinsights-agent/*/*.pom'
```
Comment (trask) [inline .pipelines/pipeline.user.windows.official.yml:26]: @heyams I'm still debugging smoke test failure, but can you take a look at the changes I needed to make to pipelines b/c sources/javadocs/pom are no longer built under build/libs, but rather only built using `publishToMavenLocal` under ~/.m2

File: .pipelines/pipeline.user.windows.buddy.yml [hunk 7]
```diff
@@ -13,9 +13,4 @@ build:
       artifacts:
         - to: 'Artifacts'
           include:
-            - '**/build/libs/*.jar'
-            - '**/build/libs/*-sources.jar'
-            - '**/build/libs/*-javadoc.jar'
-            - '**/build/libs/*.pom'
-          exclude:
-            - 'buildSrc/**/*'
+            - 'agent/agent/build/libs/*.jar'
```
Comment (trask) [inline .pipelines/pipeline.user.windows.buddy.yml:16]: @heyams

File: .pipelines/pipeline.user.windows.official.nonrelease.yml [hunk 8]
```diff
@@ -20,9 +20,4 @@ build:
       artifacts:
         - to: 'Artifacts'
           include:
-            - '**/build/libs/*.jar'
-            - '**/build/libs/*-sources.jar'
-            - '**/build/libs/*-javadoc.jar'
-            - '**/build/libs/*.pom'
-          exclude:
-            - 'buildSrc/**/*'
+            - 'agent/agent/build/libs/*.jar'
```
Comment (trask) [inline .pipelines/pipeline.user.windows.official.nonrelease.yml:23]: @heyams

File: .pipelines/pipeline.user.windows.official.yml [hunk 9]
```diff
@@ -16,13 +16,11 @@ build:
     - !!buildcommand
       name: 'Assemble ApplicationInsights-Java JARs'
       command: '.scripts/gradle.cmd'
-      arguments: 'assemble copyLibsToGlobalArtifactsFolder -DisRelease=true -Pai.etw.native.build=release'
+      arguments: 'publishToMavenLocal -DisRelease=true -Pai.etw.native.build=release'
```
Comment (heyams) [inline .pipelines/pipeline.user.windows.official.yml:19]: why do you want to use 'publishToMavenLocal' for builds? can we keep the old one? what is the reason of this change?

File: .pipelines/pipeline.user.windows.official.yml (hunk 6 above)
Comment (heyams) [reply .pipelines/pipeline.user.windows.official.yml:26]: image

all these yml(s) need to be updated.

Comment (heyams) [reply .pipelines/pipeline.user.windows.official.yml:26]: release pipeline needs to to be updated too. jarPath is different with this change.

File: .pipelines/pipeline.user.windows.official.yml (hunk 9 above)
Comment (trask) [reply .pipelines/pipeline.user.windows.official.yml:19]: publishToMavenLocal is the standard way to build javadoc/sources/pom, what we were doing previously was non-standard

File: .pipelines/pipeline.user.windows.official.yml (hunk 6 above)
Comment (trask) [reply .pipelines/pipeline.user.windows.official.yml:26]: jar is still built to same place, just the source/javadoc/pom file no longer get build there

File: .pipelines/pipeline.user.windows.buddy.yml (hunk 7 above)
Comment (heyams) [reply .pipelines/pipeline.user.windows.buddy.yml:16]: this is fine without including pom, sources, and javadoc.. that's for maven publish.

File: .pipelines/pipeline.user.windows.official.nonrelease.yml (hunk 8 above)
Comment (heyams) [reply .pipelines/pipeline.user.windows.official.nonrelease.yml:23]: this is fine without including pom, sources, and javadoc..

File: .pipelines/pipeline.user.windows.official.yml (hunk 6 above)
Comment (heyams) [reply .pipelines/pipeline.user.windows.official.yml:26]: looks good to me on the yml.. I will update later once you merge if needed.

**Comment by heyams** [comment]
copyBitsToAzureBlob.ps1 will fail due to source, and pom path change.
i will update it later once you merge this PR. need to pass two paths now.. agent jar and then the new path for the source, pom and javadoc.

**Comment by heyams** [comment]
copyBitsToAzureBlob.ps1 is used in the release pipeline.
